### DISCUSSION TITLE
Half-hour offset

### DISCUSSION THREAD

**Discussion by majiiidh** [martinet101/ElevenClock#1255] [General]
Is it possible to add a half-hour offset?
The custom formats of {%H+1.5} or {%M+30} are not working.

**Comment by marticliment** [comment]
This is not possible to do right now.

Please open a feature request if you wish to be able to do so

**Comment by arshanskiyav** [comment]
> Is it possible to add a half-hour offset? The custom formats of {%H+1.5} or {%M+30} are not working.

Hi I have the same problem and my solution is to replace in tools.py:

```
def evaluate_simple_expression(expression: str):  # supported expressions are of form x +/- y
        tokens = re.split(r'(\+|\-)', expression)
        result = int(tokens[0].strip())

        for i in range(1, len(tokens), 2):
            operator = tokens[i].strip()
            operand = int(tokens[i + 1].strip())

            if operator == "+":
                result += operand
            elif operator == "-":
                result -= operand

        return result
```

to
```
def evaluate_simple_expression(expression: str):  # supported expressions are of form x +/- y
        tokens = re.split(r'(\+|\-)', expression)
        result = int(tokens[0].strip())

        for i in range(1, len(tokens), 2):
            operator = tokens[i].strip()
            operand = int(tokens[i + 1].strip())

            if operator == "+":
                result += operand
            elif operator == "-":
                result -= operand

            if result >= 60 and operand > 15:
               result -= 60
            if result >= 24 and operand < 15:
               result -= 24

        return result
```

Yes, this solution is not ideal, but minutes and hours works fine.
The maximum offset for the time zone is 14 hours, the minimum step in minutes according to the technical specifications is 15 minutes

On picture example GMT+1.5
![2023-11-02_17-33-14](https://github.com/marticliment/ElevenClock/assets/13488664/6fe52422-f53e-4de4-9b5f-78404ea5e281)

**Comment by arshanskiyav** [comment]
> > Is it possible to add a half-hour offset? The custom formats of {%H+1.5} or {%M+30} are not working.
> 
> Hi I have the same problem and my solution is to replace in tools.py:
> 
> ```
> def evaluate_simple_expression(expression: str):  # supported expressions are of form x +/- y
>         tokens = re.split(r'(\+|\-)', expression)
>         result = int(tokens[0].strip())
> 
>         for i in range(1, len(tokens), 2):
>             operator = tokens[i].strip()
>             operand = int(tokens[i + 1].strip())
> 
>             if operator == "+":
>                 result += operand
>             elif operator == "-":
>                 result -= operand
> 
>         return result
> ```
> 
> to
> 
> ```
> def evaluate_simple_expression(expression: str):  # supported expressions are of form x +/- y
>         tokens = re.split(r'(\+|\-)', expression)
>         result = int(tokens[0].strip())
> 
>         for i in range(1, len(tokens), 2):
>             operator = tokens[i].strip()
>             operand = int(tokens[i + 1].strip())
> 
>             if operator == "+":
>                 result += operand
>             elif operator == "-":
>                 result -= operand
> 
>             if result >= 60 and operand > 15:
>                result -= 60
>             if result >= 24 and operand < 15:
>                result -= 24
> 
>         return result
> ```
> 
> Yes, this solution is not ideal, but minutes and hours works fine. The maximum offset for the time zone is 14 hours, the minimum step in minutes according to the technical specifications is 15 minutes
> 
> On picture example GMT+1.5 ![2023-11-02_17-33-14](https://user-images.githubusercontent.com/13488664/280062460-6fe52422-f53e-4de4-9b5f-78404ea5e281.png)

Sorry I made a mistake,  if time is 21:35 and +1.5 hours and we use pattern {%H+1}:{%M+30}, then we get wrong value - 22:05 (((
Therefore, this solution cannot be used to offset by half an hour.

**Reply by marticliment** [reply]
@arshanskiyav, could you (only if you want to) please create a PR with your changes?

**Comment by arshanskiyav** [comment]
Hi,
Hi, I created a new version with the numbers added correctly, but I don't
have permission to PR

Sending refs/heads/add_offsets
Remote: Permission to marticliment/ElevenClock.git denied to arshanskiyav.
Git failed with a fatal error.
unable to access 'https://github.com/marticliment/ElevenClock.git/': The
requested URL returned error: 403


чт, 2 нояб. 2023 г. в 21:02, Martí Climent ***@***.***>:

> @arshanskiyav <https://github.com/arshanskiyav>, could you (only if you
> want to) please create a PR with your changes?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/marticliment/ElevenClock/discussions/1255#discussioncomment-7460701>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADG5EGDVPSTI423WNW6WDH3YCP335AVCNFSM6AAAAAA4XX53O2VHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TINRQG4YDC>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 
*С Уважением, *
*Аршанский Андрей Владимирович*

*+7(950) 014-24-46*
***@***.*** ***@***.***>*

**Reply by marticliment** [reply]
please try creating the pr through github web

**Comment by arshanskiyav** [comment]
Done

пн, 6 нояб. 2023 г. в 15:23, Martí Climent ***@***.***>:

> please try creating the pr through github web
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/marticliment/ElevenClock/discussions/1255#discussioncomment-7488198>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADG5EGARB3EIGYHCXRYWDOLYDDXFVAVCNFSM6AAAAAA4XX53O2VHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TIOBYGE4TQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 
*С Уважением, *
*Аршанский Андрей Владимирович*

*+7(950) 014-24-46*
***@***.*** ***@***.***>*

**Comment by arshanskiyav** [comment]
> @arshanskiyav, could you (only if you want to) please create a PR with your changes?



> Is it possible to add a half-hour offset? The custom formats of {%H+1.5} or {%M+30} are not working.

Hi, in the new version I have added a function to add a half-hour offset.
Could you test this?
Please read the instruction before use

**Reply by marticliment** [reply]
the code looks fine, initial testing works.

I will do further testing soon

**Reply by mjhabib** [reply]
Hi, thank you for your work. I just wanted to remind you that the problem mentioned in this topic still exists. I’m in the process of testing this software, and the only feature I needed was to show a different clock on my taskbar.
I achieved this goal by following your suggestion in this topic [1152](https://github.com/marticliment/ElevenClock/discussions/1152), despite a few bugs which are not related to this topic.
Now, for the sake of this topic, when I add a half-hour offset to the clock, if, for example, the time is 10:02, it shows the time as 10:62, which I think [arshanskiyav](https://github.com/arshanskiyav) provides a solution for, but it has not been addressed yet.
I appreciate your time.

**Reply by arshanskiyav** [reply]
Hi,
check the new version
![image](https://github.com/marticliment/ElevenClock/assets/13488664/154742e5-f147-4287-8af2-8e0639c95de1)
![image](https://github.com/marticliment/ElevenClock/assets/13488664/ef224bd7-0fa0-4cac-b84c-b4262bfc7c33)
![image](https://github.com/marticliment/ElevenClock/assets/13488664/b5a12b74-ad9f-425e-90ae-dbde600ece55)
![image](https://github.com/marticliment/ElevenClock/assets/13488664/f94dac30-2c3c-473e-9527-ea27d6511d27)

**Reply by mjhabib** [reply]
Thank you sir, I wasn't aware of this format.

**Reply by arshanskiyav** [reply]
This format was added in version 4.3.3 on November 25, 2023.
![image](https://github.com/marticliment/ElevenClock/assets/13488664/df0f97cf-4d3e-4a8b-9639-1922a6552574)
